### PR TITLE
Reduce runtime panics through `SystemParam` validation

### LINKED ISSUE #15265: Systems should be skipped if their resources cannot be fetched
> The solution to this I would prefer is to lean into the ECS paradigm more & treat more things as if they were querying the world. If a Res<Foo> doesn't exist the system just doesn't run. Similar to how iterating over a Query<Bar> that matches nothing loops over nothing. If users want a more explicit panicky behavior they can use Option<Res<Foo>> & unwrap.

I do think we should do this, and if we can get this behavior in place I'd prefer to encourage the use of the `Single` system param from #15264. This would go a long way to addressing the problem and ensuring non-panicking behavior by default without an ergonomics hit.

_Originally posted by @iiYese and @alice-i-cecile in https://github.com/bevyengine/bevy/issues/14275#issuecomment-2356344841_

### PR DESCRIPTION
**Pull request by MiniaczQ**
# Objective

The goal of this PR is to introduce `SystemParam` validation in order to reduce runtime panics.

Fixes #15265

## Solution

`SystemParam` now has a new method `validate_param(...) -> bool`, which takes immutable variants of `get_param` arguments. The returned value indicates whether the parameter can be acquired from the world. If parameters cannot be acquired for a system, it won't be executed, similarly to run conditions. This reduces panics when using params like `Res`, `ResMut`, etc. as well as allows for new, ergonomic params like #15264 or #15302.

Param validation happens at the level of executors. All validation happens directly before executing a system, in case of normal systems they are skipped, in case of conditions they return false.

Warning about system skipping is primitive and subject to change in subsequent PRs.

## Testing

Two executor tests check that all executors:
- skip systems which have invalid parameters:
  - piped systems get skipped together,
  - dependent systems still run correctly,
- skip systems with invalid run conditions:
  - system conditions have invalid parameters,
  - system set conditions have invalid parameters.

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
The opinions of the SME-ECS (@JoJoJet @maniwani @james7132 and myself) and frankly the rest of the ECS community are very positive on the broad direction of this work, so I'm marking it as `X-Blessed`.

**Comment by MiniaczQ** [comment]
`validate_param` is technically not `unsafe`, but we might want to mark it as so.
In theory, this method shouldn't access components, despite having a readonly world access.

File: crates/bevy_ecs/src/system/system.rs
Comment (urben1680) [inline crates/bevy_ecs/src/system/system.rs:102]:
```suggestion
    /// NOTE: The tick from [`World::read_change_tick`] isn't the
    /// exact same as the one that will be passed to `run_unsafe`,
    /// none the less it should be enough to check monothicity.
```

Comment (MiniaczQ) [reply crates/bevy_ecs/src/system/system.rs:102]: Oops, that was a leftover comment, we don't pass tick to `validate_param` anymore

File: crates/bevy_ecs/src/schedule/executor/multi_threaded.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:533]: This needs to be resolved before merging.

File: crates/bevy_ecs/src/schedule/executor/multi_threaded.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:531]: We should also clean this up before merging.

**Review by alice-i-cecile — COMMENTED** [review]
Very pleased about how non-intrusive this is!

File: crates/bevy_ecs/src/schedule/executor/multi_threaded.rs
Comment (MiniaczQ) [reply crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:533]: Moving validation or warning

File: crates/bevy_ecs/src/schedule/executor/multi_threaded.rs
Comment (MiniaczQ) [reply crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:531]: This will be pretty intrusive to the existing executors, not sure if it's the right step, I'd like some feedback here if possible

File: crates/bevy_ecs/src/schedule/executor/multi_threaded.rs
Comment (MiniaczQ) [reply crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:533]: How much of the warning mechanic do we want here?
`warn_once` is just wrong,
`warn` will run every time
and the goal is a `warn_once` with system identifier as argument

Comment (urben1680) [reply crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:533]: We could store a Set in the world to keep track about which system we warned about.

File: crates/bevy_ecs/src/schedule/executor/multi_threaded.rs
Comment (alice-i-cecile) [reply crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:531]: Right, so the only benefit is saving some checks right?

File: crates/bevy_ecs/src/schedule/executor/multi_threaded.rs
Comment (MiniaczQ) [reply crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:533]: Accessing world component data is a major no here,
I'll probably make a rw hashset static which stores system identifiers

Comment (alice-i-cecile) [reply crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:533]: I'd accept just `warn` for an initial merge. `warn_once` is just wrong. Before 0.15, I'd like to get this swapped to a warn-once-per-system + configurable warnings. I don't really care if that's done globally via an atomic hashset or something, or via a set on the World.

Comment (urben1680) [reply crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:533]: >  I'll probably make a rw hashset static which stores system identifiers

That is not specific to an instance of World/App then but I guess that'd be fine with just warnings.

Comment (alice-i-cecile) [reply crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:533]: Yeah, a static is basically what we use for `warn_once` :)

File: crates/bevy_ecs/src/schedule/executor/multi_threaded.rs
Comment (MiniaczQ) [reply crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:531]: From discussion around here https://discord.com/channels/691052431525675048/749335865876021248/1286033698327892018
we're fully giving up on short circuiting, so I can remove those todos

**Review by alice-i-cecile — APPROVED** [review]
As follow-up, we should make the warning behavior configurable, as discussed in #15625.

**Comment by MiniaczQ** [comment]
Blocked on #15289 (if we want miri resolved)
Also needs D-Unsafe tag

**Comment by Victoronz** [comment]
Blocked on #15309, #15289 is no longer relevant

**Comment by iiYese** [comment]
I'll decline to review, too deep in the ECS internals for my experience. Can't say anything confidently.

**Review by chescock — APPROVED** [review]

File: crates/bevy_ecs/src/system/system_param.rs
Comment (alice-i-cecile) [reply crates/bevy_ecs/src/system/system_param.rs:2324]: Agreed, I think this is a good default.

**Comment by alice-i-cecile** [comment]
@MiniaczQ I'd like to get these last few cleanup items done, but I know you're busy. Let me know if you'd like help, or for us to just merge this as is and punt the rest to follow-up.

File: crates/bevy_ecs/src/system/system_param.rs
Comment (MiniaczQ) [reply crates/bevy_ecs/src/system/system_param.rs:2324]: Defaulting to `true` will cause at best worst case panics, at worst worst case UB.
I'd rather every manual implementation give this a thought.

File: crates/bevy_render/src/extract_param.rs
Comment (MiniaczQ) [reply crates/bevy_render/src/extract_param.rs]: Added, I'd like someone to make sure this is correct, not super familiar with multi-world stuff

File: crates/bevy_ecs/src/system/system_param.rs
Comment (MiniaczQ) [reply crates/bevy_ecs/src/system/system_param.rs:2320]: Can we merge this as is that possibly default `false` in another PR?

File: crates/bevy_render/src/extract_param.rs
Comment (chescock) [reply crates/bevy_render/src/extract_param.rs]: Hmm, this might actually be tricky.  It looks like `get_param` passes the *main* world into `state.state.get`, so you should probably pass that instead of `world`.  But the only way to get that is `Res::<MainWorld>::get_param`, which requires `&mut Self::State` instead of just `&Self::State`.  

Maybe you should put it back to what you had originally and make a follow-up issue to make `validate_param` work with `Extract` parameters.

Comment (MiniaczQ) [reply crates/bevy_render/src/extract_param.rs]: I'll mark it as TODO and provide default impl of `true`

**Review by chescock — APPROVED** [review]

File: crates/bevy_ecs/src/system/system_param.rs
Comment (hymm) [inline crates/bevy_ecs/src/system/system_param.rs:228]: Do we have docs anywhere that the implementor is only allowed to read the registered data + world metadata and not allowed to mutate any data through the `UnsafeWorldCell`

**Review by hymm — COMMENTED** [review]
I made some comments on the safety for the single_threaded executor and I believe the simple executor has the same issues. It's late here so I might be missing something obvious, but I don't see `update_archetype_component_access` being called for the conditions for either of those two executors.

File: crates/bevy_ecs/src/schedule/executor/single_threaded.rs
Comment (MiniaczQ) [reply crates/bevy_ecs/src/schedule/executor/single_threaded.rs:94]: This is the same requirement for `get_param` which happens inside every `System::run_unsafe`,
since `validate_param` is essentially `get_param` with read-only restrictions, it inherits the same safety.
I have no clue why it's never called for single threaded executors, I raised this in discord but didn't get attention

File: crates/bevy_ecs/src/schedule/executor/single_threaded.rs
Comment (MiniaczQ) [reply crates/bevy_ecs/src/schedule/executor/single_threaded.rs:92]: I see your point, should we special case single-threaded comments though?
Seems like more maintenance, but I can change it

File: crates/bevy_ecs/src/schedule/executor/single_threaded.rs
Comment (MiniaczQ) [reply crates/bevy_ecs/src/schedule/executor/single_threaded.rs:182]: answered in another reply

File: crates/bevy_ecs/src/schedule/executor/single_threaded.rs
Comment (MiniaczQ) [reply crates/bevy_ecs/src/schedule/executor/single_threaded.rs:180]: ditto

File: crates/bevy_ecs/src/system/system_param.rs
Comment (MiniaczQ) [reply crates/bevy_ecs/src/system/system_param.rs:228]: No, I'll add that

File: crates/bevy_ecs/src/schedule/executor/single_threaded.rs
Comment (chescock) [reply crates/bevy_ecs/src/schedule/executor/single_threaded.rs:94]: Oh, I remember this: `update_archetypes_component_access()` gets called by `System::run()`.  The single-threaded executors have `&mut World`, so they can call `run()` instead of `run_unsafe()` and have this handled automatically.  

Hmm, should `System::run()` check `validate_params_unsafe()`?  That would make this change work with `World::run_system()` and friends, and it would remove the need for changes to the simple executor.  

Another option is a safe `validate_params()` that takes `&mut World` and calls `update_archetype_component_access()` itself.

Comment (maniwani) [reply crates/bevy_ecs/src/schedule/executor/single_threaded.rs:94]: > I have no clue why `update_archetype_component_access` is never called for single threaded executors, I raised this in Discord but didn't get attention.

Were you referring to `main` here?
On `main`, the simple/single-threaded executors call the safe `run` function for all systems, including conditions.

https://github.com/bevyengine/bevy/blob/e3b6b125a0a3b9bd15d939b51136cc186f086c56/crates/bevy_ecs/src/system/system.rs#L77-L86

Comment (MiniaczQ) [reply crates/bevy_ecs/src/schedule/executor/single_threaded.rs:94]: okay, this complicates `validate_param`

Comment (MiniaczQ) [reply crates/bevy_ecs/src/schedule/executor/single_threaded.rs:94]: Moving this to discord discussion
https://discord.com/channels/691052431525675048/1287411721593426031

Comment (MiniaczQ) [reply crates/bevy_ecs/src/schedule/executor/single_threaded.rs:94]: Agreed on duplicate `update_archetypes_component_access` for safe variant of `validate_param`

File: crates/bevy_ecs/src/schedule/executor/single_threaded.rs
Comment (MiniaczQ) [reply crates/bevy_ecs/src/schedule/executor/single_threaded.rs]: No longer important, single-threaded executors use safe `validate_param`

File: crates/bevy_ecs/src/system/system_param.rs
Comment (MiniaczQ) [reply crates/bevy_ecs/src/system/system_param.rs:2189]: done here

File: crates/bevy_render/src/extract_param.rs
Comment (MiniaczQ) [reply crates/bevy_render/src/extract_param.rs]: implemented

File: crates/bevy_ecs/src/system/function_system.rs
Comment (SkiFire13) [inline crates/bevy_ecs/src/system/function_system.rs:348]: Shouldn't this mention `archetype_component_access` instead?

File: crates/bevy_ecs/src/system/system.rs
Comment (SkiFire13) [inline crates/bevy_ecs/src/system/system.rs:114]: Is there a reason this methods takes a `&mut World` but only uses it in a read-only way?

File: crates/bevy_ecs/src/system/system.rs
Comment (SkiFire13) [inline crates/bevy_ecs/src/system/system.rs:103]: Why does this mention `update_archetype_component_access` while the other `validate_param`s mention `init_state`?

File: crates/bevy_ecs/src/system/system_param.rs
Comment (SkiFire13) [inline crates/bevy_ecs/src/system/system_param.rs:1663]: Shouldn't this return `true` since `ParamSet::get_param` cannot panic? Or does this want to prevent panics when calling `.p0()`, `.p1()` etc etc? Though the user might not call them when they would panic, so this would silently change the behaviour in those cases.

File: crates/bevy_ecs/src/world/mod.rs
Comment (SkiFire13) [inline crates/bevy_ecs/src/world/mod.rs:1491]: Nit: this could use `is_some_and`

File: crates/bevy_ecs/macros/src/lib.rs
Comment (SkiFire13) [inline crates/bevy_ecs/macros/src/lib.rs:278]:
```suggestion
                    world: UnsafeWorldCell<'w>,
```

File: crates/bevy_ecs/macros/src/lib.rs
Comment (SkiFire13) [inline crates/bevy_ecs/macros/src/lib.rs:280]: Is there a reason this doesn't use full paths, but the other `validate_param` in this file does?

File: crates/bevy_ecs/src/schedule/executor/multi_threaded.rs
Comment (SkiFire13) [inline crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:574]: This SAFETY comment of this function only mentions run conditions, it should be updated to reflect this requirement.

File: crates/bevy_ecs/src/system/system.rs
Comment (SkiFire13) [inline crates/bevy_ecs/src/system/system.rs:98]: This should however mention that the implementation of `System` should not rely on this fact being respected for safety. That is, safe code can still call `System::run` even if `System::validate_param` return `false` and that should panic rather than causing UB.

File: crates/bevy_ecs/src/system/system_param.rs
Comment (SkiFire13) [inline crates/bevy_ecs/src/system/system_param.rs:224]: Similarly this should mention that `get_param` can get called even if `validate_param` returns `false`, and in that it should panic rather than resulting in UB.

File: crates/bevy_ecs/src/system/system_param.rs
Comment (MiniaczQ) [reply crates/bevy_ecs/src/system/system_param.rs:1663]: If any parts of the set are invalid, the system won't run.
Panicking per variant is something to be considered when systems can return `Result` and we can use `?`.

File: crates/bevy_ecs/macros/src/lib.rs
Comment (MiniaczQ) [reply crates/bevy_ecs/macros/src/lib.rs:280]: Consistency with the rest of the macro

File: crates/bevy_ecs/src/system/system_param.rs
Comment (MiniaczQ) [reply crates/bevy_ecs/src/system/system_param.rs:1663]: We can remove validation from param set and keep the panicking API for now
@alice-i-cecile?

File: crates/bevy_ecs/src/system/system.rs
Comment (MiniaczQ) [reply crates/bevy_ecs/src/system/system.rs:107]: This was copy-paste from `get_param`/`run` methods, this means that they are also inconsistent with each other.
Agreed it should refer to `update_archetype_component_access` and some reference to `new_archetype` instead.

File: crates/bevy_ecs/src/system/system_param.rs
Comment (alice-i-cecile) [reply crates/bevy_ecs/src/system/system_param.rs:1666]: Let's cut it from this PR and keep the panicking API. `ParamSet` is niche and that's easy follow-up.

Comment (MiniaczQ) [reply crates/bevy_ecs/src/system/system_param.rs:1666]: I'll consider this a yes, the change itself is trivial

File: crates/bevy_ecs/src/system/system.rs
Comment (MiniaczQ) [reply crates/bevy_ecs/src/system/system.rs]: Addressed, let me know if that's sufficient.
If not I'm afraid we'll need a follow up PR since Alice wants this merged in the next train

Comment (SkiFire13) [reply crates/bevy_ecs/src/system/system.rs]: I guess it's good enough now, I'm ok with merging it and maybe improve the wording in a follow up PR

**Comment by alice-i-cecile** [comment]
Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to https://github.com/bevyengine/bevy-website/issues/1691 if you'd like to help out.

**Pull request merged**
